### PR TITLE
Switch over to a regular expression package that claims to be Perl 5 compatible

### PR DESCRIPTION
**Pull request by xyproto**
Regular expressions in ECMAScript 5 are modelled after the regular expression facility in the Perl 5 programming language ([ECMA 5 spec](http://www.ecma-international.org/ecma-262/5.1/#sec-15.10)).

This pull requests switches over from the `regexp` package to [P5R](https://github.com/xyproto/p5r), which is compatible with Perl 5 regular expressions.

**More testing is needed** in order to confirm for sure that Perl 5 regular expressions will be supported correctly for all cases, but it's a start.

For some functions that are supported by `regexp` but not yet by [P5R](https://github.com/xyproto/p5r), the regular expression struct is converted to a `regexp` struct and the same method call as was used before is being used. This might have a small performance hit (since the regular expression string is recompiled), but from my initial testing it should not have too large of an impact.

In the future, it will probably be better to use the functions from P5R instead, which takes different parameters (and often can return an error where `regexp` returns none). I have added comments in the code where I think this might be a good idea.

Hopefully this pull request can bring Otto one step closer to becoming ECMAScript 5 compatible, which would be an advantage.

### CODE REVIEW

**Comment by xyproto** [comment]
All tests pass when running `go test`.

File: builtin.go [hunk 1]
```diff
@@ -217,8 +217,9 @@ func _builtinGlobal_encodeURI(call FunctionCall, escape *regexp.Regexp) Value {
 		output = append(output, encode...)
 	}
 	{
-		value := escape.ReplaceAllFunc(output, func(target []byte) []byte {
-			// Probably a better way of doing this
+		// FIX: Find a better way of doing this
```
Comment (stevenh) [inline builtin.go]: Indentation issue here.

**Comment by xyproto** [comment]
Moved the comment. All indentation looks okay when I run `go fmt`.

**Comment by stevenh** [comment]
Thanks for doing this @xyproto would be really good if we could be 100% compatible with JS RE's and this seems like a step in the right direction.

Do you have an benchmarks for performance of P5R vs regexp?

**Comment by xyproto** [comment]
I have the beginnings of running the [regex-dna](https://benchmarksgame.alioth.debian.org/u64q/program.php?test=regexdna&lang=node&id=2) benchmark here (attaching tarball), but I'm having problems convincing otto to read in the input file.
[regex-dna-benchmark.zip](https://github.com/robertkrimen/otto/files/331824/regex-dna-benchmark.zip).

**Comment by uppfinnarn** [comment]
There are a handful of places in [parser/regexp.go](https://github.com/xyproto/otto/blob/master/parser/regexp.go) that still rejects backreferences and lookaheads. If you could tear that out, I could do some further testing.

**Comment by xyproto** [comment]
Thanks for testing! 

Is the problem that the lack of backreferences and lookaheads in parser/regexp.go triggers errors? Are you thinking tear out as in moving that specific code to another package, or something else?

**Comment by uppfinnarn** [comment]
Oh, as in removing the errors - those checks are there to throw compiler errors rather than runtime errors for invalid regexes, but now that they're valid, there's not much of a point anymore.

**Comment by nwidger** [comment]
I don't know if this is helpful or not, but I tried running the TypeScript compiler downloaded from [here](https://raw.githubusercontent.com/Microsoft/TypeScript/master/lib/typescriptServices.js) through the xyproto/otto master branch.  It choked on a handful of regexps defined in that file.  Here's my test program and a listing of the regexps it didn't like:

``` go
package main

import (
	"io/ioutil"
	"log"

	"github.com/xyproto/otto"
)

func main() {
	buf, err := ioutil.ReadFile("typescriptServices.js")
	if err != nil {
		log.Fatalln(err)
	}

	_, err = otto.New().Run(string(buf))
	if err != nil {
		log.Fatalln(err)
	}
}
```

Running this spit out errors:

```
Invalid regular expression: re2: Invalid \N <backreference>
Invalid regular expression: re2: Invalid (?=) <lookahead>
```

for the following lines:

``` js
5584:    ts.fullTripleSlashReferencePathRegEx = /^(\/\/\/\s*<reference\s+path\s*=\s*)('|")(.+?)\2.*?\/>/;
5585:    ts.fullTripleSlashReferenceTypeReferenceDirectiveRegEx = /^(\/\/\/\s*<reference\s+types\s*=\s*)('|")(.+?)\2.*?\/>/;
5586:    ts.fullTripleSlashAMDReferencePathRegEx = /^(\/\/\/\s*<amd-dependency\s+path\s*=\s*)('|")(.+?)\2.*?\/>/;
6675:    var isNoDefaultLibRegEx = /^(\/\/\/\s*<reference\s+no-default-lib\s*=\s*)('|")(.+?)\2\s*\/>/gim;
12905:   var amdModuleNameRegEx = /^\/\/\/\s*<amd-module\s+name\s*=\s*('|")(.+?)\1/gim;
12914:   var pathRegex = /\spath\s*=\s*('|")(.+?)\1/gim;
12915:   var nameRegex = /\sname\s*=\s*('|")(.+?)\1/gim;
45847:   var wildcardDirectoryPattern = /^[^*?]*(?=\/[^/]*[*?])/;
49128:   var cleanedTypingNames = ts.map(inferredTypingNames, function (f) { return f.replace(/((?:\.|-)min(?=\.|$))|((?:-|\.)\d+)/g, ""); });
```

Here's a copy of the `typescriptServices.js` file I used: [typescriptServices.txt](https://github.com/robertkrimen/otto/files/478775/typescriptServices.txt)

**Comment by xyproto** [comment]
I think the test is really helpful, @nwidger. I have little time to fix the regexp issues right now because of the latest addition to the family, but might be able to in a few months time. Cheers.

**Comment by xvyzk** [comment]
Probably a good way to do this is to make another engine in a similar vein to the .NET/p5r one, but targeting JavaScript directly. Then integrate that.

File: builtin.go [hunk 2]
```diff
@@ -218,7 +218,7 @@ func _builtinGlobal_encodeURI(call FunctionCall, escape *regexp.Regexp) Value {
 	}
 	{
 		value := escape.ReplaceAllFunc(output, func(target []byte) []byte {
-			// Probably a better way of doing this
+			// TODO There is probably a better way of doing this
```
Comment (stevenh) [inline builtin.go]: Please use:
```go
// TODO: There is ...
```

File: otto/main.go [hunk 3]
```diff
@@ -6,8 +6,8 @@ import (
 	"io/ioutil"
 	"os"
 
-	"github.com/robertkrimen/otto"
 	"github.com/robertkrimen/otto/underscore"
+	"github.com/xyproto/otto"
```
Comment (stevenh) [inline otto/main.go]: This looks like a miss merge

**Comment by xyproto** [comment]
I want to prioritize other projects and abandon my fork of otto. Please fork the fork if you're interested. Cheers.

**Pull request closed without merging**